File: src/Core/FinancialEvents/Services/FinancialEventsProcessor.py
from typing import Sequence

import Core.FinancialEvents.Schemas.Grouping as pgf
from Core.FinancialEvents.Schemas.LotMatchingConfiguration import (
    LotMatchingConfiguration,
)
from Core.FinancialEvents.Utils.ProcessingUtils import ProcessingUtils
from Core.LotMatching.Services.LotMatcher import LotMatcher


class FinancialEventsProcessor:
    """
    Note that since the LotMatcher is stateful, we need to create a new instance for each financial grouping.
    """

    def __init__(
        self,
        processingUtils: ProcessingUtils,
        lotMatcher: LotMatcher,
    ) -> None:
        self.lotMatcher = lotMatcher
        self.processingUtils = processingUtils

    def process(
        self, input: pgf.FinancialGrouping, lotMatchingConfiguration: LotMatchingConfiguration
    ) -> pgf.UnderlyingGroupingWithTradesOfInterest:
        lotMatcher = self.lotMatcher

        lotMatchingMethodInstance = lotMatchingConfiguration.ForStocks(input)
        stockTradesOfInterest = lotMatcher.matchLotsWithGenericTradeEvents(lotMatchingMethodInstance, input.StockTrades)

        stockTradesOfInterestFiltered = stockTradesOfInterest.getTradesOfLotsClosedInPeriod(
            periodStart=lotMatchingConfiguration.fromDate, periodEnd=lotMatchingConfiguration.toDate
        )

        derivativeGroupingsOfInterest: list[pgf.UnderlyingDerivativeGrouping] = []
        for derivativeGrouping in input.DerivativeGroupings:
            lotMatchingMethodInstance = lotMatchingConfiguration.ForDerivatives(input)
            derivativeTradesOfInterest = lotMatcher.matchLotsWithGenericTradeEvents(
                lotMatchingMethodInstance, derivativeGrouping.DerivativeTrades
            )

            derivativeTradesOfInterestFiltered = derivativeTradesOfInterest.getTradesOfLotsClosedInPeriod(
                periodStart=lotMatchingConfiguration.fromDate, periodEnd=lotMatchingConfiguration.toDate
            )

            derivativeGroupingsOfInterest.append(
                pgf.UnderlyingDerivativeGrouping(
                    FinancialIdentifier=derivativeGrouping.FinancialIdentifier,
                    DerivativeTrades=derivativeTradesOfInterestFiltered.Trades,  # TODO: Types
                )
            )

        interestingGrouping = pgf.UnderlyingGroupingWithTradesOfInterest(
            FinancialIdentifier=input.FinancialIdentifier,
            CountryOfOrigin=input.CountryOfOrigin,
            UnderlyingCategory=input.UnderlyingCategory,
            StockTrades=stockTradesOfInterestFiltered.Trades,  # TODO: Types
            DerivativeGroupings=derivativeGroupingsOfInterest,
            CashTransactions=input.CashTransactions,
        )

        return interestingGrouping

    def generateInterestingUnderlyingGroupings(
        self, groupings: Sequence[pgf.FinancialGrouping], lotMatchingMethod: LotMatchingConfiguration
    ) -> Sequence[pgf.UnderlyingGroupingWithTradesOfInterest]:
        processedGroupings = list(map(lambda grouping: self.process(grouping, lotMatchingMethod), groupings))
        return processedGroupings

File: lib/Core/FinancialEvents/FinancialEventsProcessor.ts
import type { FinancialGrouping } from "@brrr/Core/Schemas/Grouping.ts";
import { UnderlyingDerivativeGrouping, UnderlyingGroupingWithTradesOfInterest } from "@brrr/Core/Schemas/Grouping.ts";
import type { LotMatchingConfiguration } from "@brrr/Core/Schemas/LotMatchingConfiguration.ts";
import type { LotMatcher } from "@brrr/Core/LotMatching/LotMatcher.ts";
import type { LotMatchingMethod } from "@brrr/Core/LotMatching/LotMatchingMethod.ts";

export class FinancialEventsProcessor {
	private lotMatcher: LotMatcher;

	constructor(_processingUtils: unknown, lotMatcher: LotMatcher) {
		this.lotMatcher = lotMatcher;
	}

	// Note: since the LotMatcher is stateful, a new LotMatchingMethod instance must be created
	// for each financial grouping (done via lotMatchingConfiguration.forStocks/forDerivatives).
	process(input: FinancialGrouping, lotMatchingConfiguration: LotMatchingConfiguration): UnderlyingGroupingWithTradesOfInterest {
		const lotMatcher = this.lotMatcher;

		const lotMatchingMethodInstance = lotMatchingConfiguration.forStocks(input) as unknown as LotMatchingMethod;
		const stockTradesOfInterest = lotMatcher.matchLotsWithGenericTradeEvents(lotMatchingMethodInstance, input.stockTrades);

		const stockTradesOfInterestFiltered = stockTradesOfInterest.getTradesOfLotsClosedInPeriod(
			lotMatchingConfiguration.fromDate,
			lotMatchingConfiguration.toDate,
		);

		const derivativeGroupingsOfInterest: UnderlyingDerivativeGrouping[] = [];
		for (const derivativeGrouping of input.derivativeGroupings) {
			const derivMethodInstance = lotMatchingConfiguration.forDerivatives(input) as unknown as LotMatchingMethod;
			const derivativeTradesOfInterest = lotMatcher.matchLotsWithGenericTradeEvents(
				derivMethodInstance,
				derivativeGrouping.derivativeTrades,
			);

			const derivativeTradesOfInterestFiltered = derivativeTradesOfInterest.getTradesOfLotsClosedInPeriod(
				lotMatchingConfiguration.fromDate,
				lotMatchingConfiguration.toDate,
			);

			derivativeGroupingsOfInterest.push(
				new UnderlyingDerivativeGrouping({
					financialIdentifier: derivativeGrouping.financialIdentifier,
					derivativeTrades: derivativeTradesOfInterestFiltered.trades as never,
				}),
			);
		}

		return new UnderlyingGroupingWithTradesOfInterest({
			financialIdentifier: input.financialIdentifier,
			countryOfOrigin: input.countryOfOrigin,
			underlyingCategory: input.underlyingCategory,
			stockTrades: stockTradesOfInterestFiltered.trades as never,
			derivativeGroupings: derivativeGroupingsOfInterest,
			cashTransactions: input.cashTransactions,
		});
	}

	generateInterestingUnderlyingGroupings(
		groupings: FinancialGrouping[],
		lotMatchingMethod: LotMatchingConfiguration,
	): UnderlyingGroupingWithTradesOfInterest[] {
		return groupings.map((g) => this.process(g, lotMatchingMethod));
	}
}
